db.go

```diff
@@ -352,22 +357,94 @@ type DB struct {
-		IgnoreUnmappedCols: true,
-		IgnoreMissingCols:  false,
-		Context:            context.Background(),
-		Logger:             nil,
-		models:             map[reflect.Type]*Model{},
-		mappings:           map[reflect.Type]fieldMap{},
+		DB:                    db,
+		AllowStringQueries:    true,
+		IgnoreUnmappedCols:    true,
+		IgnoreMissingCols:     false,
+		Context:               context.Background(),
+		Logger:                nil,
+		deadlineQueryRewriter: noopDeadlineQueryRewriter,
+		models:                map[reflect.Type]*Model{},
+		mappings:              map[reflect.Type]fieldMap{},
+	}
+}
+
+func noopDeadlineQueryRewriter(db *DB, query interface{}, millis int64) (queryWithDeadline interface{}, err error) {
+	return query, nil
+}
+
+func perconaDeadlineQueryRewriter(db *DB, query interface{}, millis int64) (queryWithDeadline interface{}, err error) {
+	serializer, err := db.getSerializer(query)
+	if err != nil {
+		return nil, err
+	}
+	querystr, err := Serialize(serializer)
+	if err != nil {
+		return nil, err
+	}
+	return fmt.Sprintf("SET STATEMENT max_statement_time=%d FOR %s", millis, querystr), nil
+}
+
+func mySql57DeadlineQueryRewriter(db *DB, query interface{}, millis int64) (queryWithDeadline interface{}, err error) {
+	serializer, err := db.getSerializer(query)
+	if err != nil {
+		return nil, err
+	}
+	querystr, err := Serialize(serializer)
+	if err != nil {
+		return nil, err
+	}
+	if strings.HasPrefix(querystr, "SELECT ") {
+		return fmt.Sprintf("SELECT /*+ MAX_EXECUTION_TIME(%d) */ %s", millis, querystr[7:]), nil
+	} else {
+		return querystr, nil
 	}
 }
 
+type QueryDeadlineType int
+
+const (
+	// When passed to WithQueryDeadlines, disable query deadlines.
+	QUERY_DEADLINES_NONE QueryDeadlineType = iota
+	// When passed to WithQueryDeadlines, enable query deadlines on Percona Server 5.6 or later.
+	QUERY_DEADLINES_PERCONA_56
+	// When passed to WithQueryDeadlines, enable query deadlines on MySql 5.7.4 or later.
+	QUERY_DEADLINES_MYSQL_57
+)
+
+// WithQueryDeadlines returns a DB instance with query deadlines enabled.
+// When the underlying database is Percona Server 5.6+, use queryDeadlineType
+// QUERY_DEADLINES_PERCONA_56. When the underlying database is MySql 5.7.4+,
+// use queryDeadlineType QUERY_DEADLINES_MYSQL_57. To disable query deadlines
+// (the default), use queryDeadlineType QUERY_DEADLINES_NONE.
+//
+// Query deadlines only affect SELECT queries run via the Executor.Query function.
+// It us up to the underlying database engine to enforce the deadline.
+func (db *DB) WithQueryDeadlines(queryDeadlineType QueryDeadlineType) *DB {
+	newDB := *db
+
+	switch queryDeadlineType {
+	case QUERY_DEADLINES_NONE:
+		newDB.deadlineQueryRewriter = noopDeadlineQueryRewriter
+	case QUERY_DEADLINES_PERCONA_56:
+		newDB.deadlineQueryRewriter = perconaDeadlineQueryRewriter
+	case QUERY_DEADLINES_MYSQL_57:
+		newDB.deadlineQueryRewriter = mySql57DeadlineQueryRewriter
+	default:
+		// Ignore unknown type
```

Comment:
```
db := db.
  WithQueryDeadlines(QueryDeadlineType(1)).
  WithQueryDeadlines(QueryDeadlineType(2))
```
produces a DB with deadline type 2.
```
db := db.
  WithQueryDeadlines(QueryDeadlineType(1)).
  WithQueryDeadlines(QueryDeadlineType(3))
```
produces a DB with deadline type 1.

Semantically, I think this behavior is inconsistent. I would expect it use deadline type 3. Granted, the DB doesn't know how to produce queries for type 3, so the safe behavior would be to use `noopDeadlineQueryRewriter` to implement it. The method could also return an error, but that complicates the API for what should be a rare occurrence.